File: lab/python/palindromos/palindromo_normalizado.py
import unicodedata


def remover_acentos(txt: str) -> str:
    nfkd = unicodedata.normalize("NFKD", txt)
    return "".join(c for c in nfkd if not unicodedata.combining(c))


def palinder(text):
    text = remover_acentos(text.strip().lower())
    original = text

    contrario = ""
    for i in range(len(text)-1, -1, -1):
        contrario += text[i]

    return contrario == original


def palinder2(text):
    text = remover_acentos(text.strip().lower())
    original = list(text)

    contrario = []
    for i in range(len(text)-1, -1, -1):
        contrario.append(text[i])

    return original == contrario


def palinder3(text):
    text = remover_acentos(text.strip().lower())
    return text == text[::-1]


print(palinder("abá"))
print(palinder2("abá"))
print(palinder2("abá"))

File: lab/javascript/palindromos/palindromo_normalizado.js
function removerAcentos(txt) {
  return txt.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
}

function palinder(text) {
  text = removerAcentos(text.trim().toLowerCase());
  let contrario = "";
  for (let i = text.length - 1; i >= 0; i--) {
    contrario += text[i];
  }
  return contrario === text;
}

function palinder2(text) {
  text = removerAcentos(text.trim().toLowerCase());
  const original = [...text];
  const contrario = [];
  for (let i = text.length - 1; i >= 0; i--) {
    contrario.push(text[i]);
  }
  return JSON.stringify(original) === JSON.stringify(contrario);
}

function palinder3(text) {
  text = removerAcentos(text.trim().toLowerCase());
  return text === [...text].reverse().join("");
}

console.log(palinder("abá"));
console.log(palinder2("abá"));
console.log(palinder3("abá"));
